docs/source/architecture.rst

```diff
@@ -1,69 +1,80 @@
 -------------------
 
-We recursively parse each of the elements, using their in built grammars.
-
-#. Initially we start with a segment, which contains only raw segments. This
-   is normally a :code:`FileSegment`, but it could in theory be any kind of
-   segment, and as we recurse it will be lower and lower sublevels.
-
-#. We then *parse* this element by calling :code:`.parse()`.
-
-   #. This first calls uses the :code:`parse_grammar` if that is present, on
-      which we call :code:`.match()`. The *match* method of this grammar will
-      return a potentially refined structure of the segments within this
-      segment in greater detail than what was initially there. In the example
-      of a :code:`FileSegment`, it first divides up the query into statements
-      and then finishes.
-
-   #. The :code:`.match()` method of any grammar is naturally recursive, and
-      is comprised of *segments* and *grammars*. The match step is still not
-      inert however and child elements of a match grammar may return more
-      specific versions of segments during this phase even though we are not
-      calling :code:`.parse()`
-
-      * *Segments* must implement a :code:`match_grammar` to be used in this
-        way. When :code:`.match()` is called on a segment, this is the grammar
-        which is used to decide whether there is a match.
-
-      * *Grammars* are objects which combine *segments* or other *grammars*
-        together in a pre-defined way. For example the :code:`OneOf` grammar
-        will match if any one of its child elements match.
+The parser is arguably the most complicated element of SQLFluff, and is
+relied on by all the other elements of the tool to do most of the heavy
+lifting.
+
+#. The lexed segments are parsed using the specified dialect's grammars. In
+   SQLFluff, grammars describe the shape of SQL statements (or their
+   components). The parser attempts to apply each potential grammar to the
+   lexed segments until all the segments have been matched.
+
+#. In SQLFluff, segments form a tree-like structure. The top-level segment is
+   a :code:`FileSegment`, which contains zero or more
+   :code:`StatementSegment`\ s, and so on. Before the segments have been parsed
+   and named according to their type, they are are 'raw', meaning they have no
+   classification other than their literal value.
+
+#. The three key components to the parser are segments, :code:`match_grammar`s
+   and :code:`parse_grammar`s. A segment can be a leaf in the parse tree, such
+   as a :code:`NumericLiteralSegment`, which is simply a number, or can contain
+   many other segments, such as a :code:`SelectStatementSegment`. Each segment
+   can specify a :code:`parse_grammar`, and a :code:`match_grammar`. The
+   :code:`match_grammar` is used to determine whether or not a given sequence
+   of segments is an instance of the segment in question. If defined, the
+   :code:`parse_grammar` is then used to match those segments further.
```

Comment:
Think we could add more detail on the distinction between match and parse grammars, with perhaps a illustrative example or two? I have done a fair amount of work on the dialects and am still unsure how I would know that I need to use both instead of just one.

Reply:
@alanmcruickshank can I rope you in here? My understanding is that match_grammars are used as a very quick way of matching a segment to tokens, vs the parse_grammar which takes longer. I'm sure there's more to it though.

Reply:
Think of the dialect like a really complicated tree diagram. Match grammars prune branches and choose which branch to follow. Parse grammars lay out which options will be available in the next round (kind of).

Reply:
Cool ok that's what I had in mind, while I'm sure it's more nuanced I think that will suffice for the docs. Really the aim of this doc is that people can read it a few times with the source code in hand and contribute to dialects and rules.

Reply:
And in the instances where there is only a match_grammar, it performs both jobs?